File: backend/anki_connect.py
"""Thin HTTP client for the AnkiConnect add-on (localhost:8765)."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any

ANKI_CONNECT_URL = "http://127.0.0.1:8765"
ANKI_CONNECT_VERSION = 6
DEFAULT_TIMEOUT_SECONDS = 3.0


class AnkiConnectError(Exception):
    """Raised when AnkiConnect is unreachable or returns an error."""


def invoke(
    action: str,
    *,
    params: dict[str, Any] | None = None,
    url: str = ANKI_CONNECT_URL,
    timeout: float = DEFAULT_TIMEOUT_SECONDS,
) -> Any:
    payload = {
        "action": action,
        "version": ANKI_CONNECT_VERSION,
        "params": params or {},
    }
    request = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            body = response.read().decode("utf-8")
    except (urllib.error.URLError, TimeoutError, OSError) as exc:
        raise AnkiConnectError(
            "AnkiConnect is unreachable. Start Anki with the AnkiConnect add-on enabled."
        ) from exc

    try:
        data = json.loads(body)
    except json.JSONDecodeError as exc:
        raise AnkiConnectError("AnkiConnect returned an invalid response.") from exc

    if not isinstance(data, dict):
        raise AnkiConnectError("AnkiConnect returned an invalid response.")

    error = data.get("error")
    if error is not None:
        raise AnkiConnectError(str(error))

    return data.get("result")


def deck_names() -> list[str]:
    result = invoke("deckNames")
    if not isinstance(result, list) or not all(isinstance(name, str) for name in result):
        raise AnkiConnectError("AnkiConnect returned an invalid deck list.")
    return result


def create_deck(deck_name: str) -> Any:
    return invoke("createDeck", params={"deck": deck_name})


def model_names() -> list[str]:
    result = invoke("modelNames")
    if not isinstance(result, list) or not all(isinstance(name, str) for name in result):
        raise AnkiConnectError("AnkiConnect returned an invalid model list.")
    return result


def model_field_names(model_name: str) -> list[str]:
    result = invoke("modelFieldNames", params={"modelName": model_name})
    if not isinstance(result, list) or not all(isinstance(name, str) for name in result):
        raise AnkiConnectError("AnkiConnect returned an invalid field list.")
    return result


def create_model(
    *,
    model_name: str,
    fields: list[str],
    card_templates: list[dict[str, str]],
    css: str | None = None,
) -> Any:
    params: dict[str, Any] = {
        "modelName": model_name,
        "inOrderFields": fields,
        "cardTemplates": card_templates,
        "isCloze": False,
    }
    if css is not None:
        params["css"] = css
    return invoke("createModel", params=params)


def add_notes(
    notes: list[dict[str, Any]],
    *,
    timeout: float = 30.0,
) -> list[Any]:
    """Create notes in Anki. Returns note IDs (null entries for failures)."""
    result = invoke("addNotes", params={"notes": notes}, timeout=timeout)
    if not isinstance(result, list):
        raise AnkiConnectError("AnkiConnect returned an invalid addNotes result.")
    return result


def _quote_anki_query_value(value: str) -> str:
    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'


def find_notes(query: str, *, timeout: float = 30.0) -> list[int]:
    result = invoke("findNotes", params={"query": query}, timeout=timeout)
    if not isinstance(result, list) or not all(isinstance(note_id, int) for note_id in result):
        raise AnkiConnectError("AnkiConnect returned an invalid findNotes result.")
    return result


def notes_info(
    note_ids: list[int],
    *,
    timeout: float = 30.0,
) -> list[dict[str, Any]]:
    if not note_ids:
        return []
    result = invoke("notesInfo", params={"notes": note_ids}, timeout=timeout)
    if not isinstance(result, list) or not all(isinstance(item, dict) for item in result):
        raise AnkiConnectError("AnkiConnect returned an invalid notesInfo result.")
    return result


def field_values_in_deck(
    deck_name: str,
    field_name: str,
    *,
    timeout: float = 30.0,
) -> set[str]:
    """Return the set of non-empty values for a note field in a deck."""
    notes = mapped_notes_in_deck(
        deck_name,
        {"value": field_name},
        timeout=timeout,
    )
    return {note["value"] for note in notes if note.get("value", "").strip() != ""}


def mapped_notes_in_deck(
    deck_name: str,
    logical_to_anki: dict[str, str],
    *,
    timeout: float = 30.0,
) -> list[dict[str, str]]:
    """Return notes in a deck with values keyed by logical field names."""
    query = f"deck:{_quote_anki_query_value(deck_name)}"
    note_ids = find_notes(query, timeout=timeout)
    notes: list[dict[str, str]] = []
    chunk_size = 250
    for start in range(0, len(note_ids), chunk_size):
        chunk = note_ids[start : start + chunk_size]
        for info in notes_info(chunk, timeout=timeout):
            fields = info.get("fields")
            if not isinstance(fields, dict):
                continue
            mapped: dict[str, str] = {}
            for logical_key, anki_field in logical_to_anki.items():
                field = fields.get(anki_field)
                if isinstance(field, dict):
                    raw = field.get("value")
                else:
                    raw = field
                mapped[logical_key] = raw.strip() if isinstance(raw, str) else ""
            notes.append(mapped)
    return notes


def is_connected() -> bool:
    try:
        deck_names()
        return True
    except AnkiConnectError:
        return False

File: frontend/src/utils/anki/ankiConnect.ts
/** Thin HTTP client for the AnkiConnect add-on (localhost:8765). */

export const ANKI_CONNECT_URL = "http://127.0.0.1:8765";
export const ANKI_CONNECT_VERSION = 6;
const DEFAULT_TIMEOUT_MS = 3000;

export class AnkiConnectError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AnkiConnectError";
  }
}

type AnkiConnectResponse = {
  result?: unknown;
  error?: string | null;
};

async function fetchWithTimeout(
  url: string,
  init: RequestInit,
  timeoutMs: number,
): Promise<Response> {
  const controller = new AbortController();
  const timer = window.setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...init, signal: controller.signal });
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      throw new AnkiConnectError(
        "AnkiConnect is unreachable. Start Anki with the AnkiConnect add-on enabled.",
      );
    }
    throw new AnkiConnectError(
      "AnkiConnect is unreachable. Start Anki with the AnkiConnect add-on enabled.",
    );
  } finally {
    window.clearTimeout(timer);
  }
}

export async function invoke(
  action: string,
  options?: {
    params?: Record<string, unknown>;
    url?: string;
    timeoutMs?: number;
  },
): Promise<unknown> {
  const url = options?.url ?? ANKI_CONNECT_URL;
  const timeoutMs = options?.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const payload = {
    action,
    version: ANKI_CONNECT_VERSION,
    params: options?.params ?? {},
  };

  let response: Response;
  try {
    response = await fetchWithTimeout(
      url,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      },
      timeoutMs,
    );
  } catch (error) {
    if (error instanceof AnkiConnectError) {
      throw error;
    }
    throw new AnkiConnectError(
      "AnkiConnect is unreachable. Start Anki with the AnkiConnect add-on enabled.",
    );
  }

  let data: AnkiConnectResponse;
  try {
    data = (await response.json()) as AnkiConnectResponse;
  } catch {
    throw new AnkiConnectError("AnkiConnect returned an invalid response.");
  }

  if (data === null || typeof data !== "object") {
    throw new AnkiConnectError("AnkiConnect returned an invalid response.");
  }

  if (data.error != null) {
    throw new AnkiConnectError(String(data.error));
  }

  return data.result;
}

export async function deckNames(): Promise<string[]> {
  const result = await invoke("deckNames");
  if (
    !Array.isArray(result) ||
    !result.every((name): name is string => typeof name === "string")
  ) {
    throw new AnkiConnectError("AnkiConnect returned an invalid deck list.");
  }
  return result;
}

export async function createDeck(deckName: string): Promise<unknown> {
  return invoke("createDeck", { params: { deck: deckName } });
}

export async function modelNames(): Promise<string[]> {
  const result = await invoke("modelNames");
  if (
    !Array.isArray(result) ||
    !result.every((name): name is string => typeof name === "string")
  ) {
    throw new AnkiConnectError("AnkiConnect returned an invalid model list.");
  }
  return result;
}

export async function modelFieldNames(modelName: string): Promise<string[]> {
  const result = await invoke("modelFieldNames", {
    params: { modelName },
  });
  if (
    !Array.isArray(result) ||
    !result.every((name): name is string => typeof name === "string")
  ) {
    throw new AnkiConnectError("AnkiConnect returned an invalid field list.");
  }
  return result;
}

export async function createModel(options: {
  modelName: string;
  fields: string[];
  cardTemplates: Array<{ Name: string; Front: string; Back: string }>;
  css?: string;
}): Promise<unknown> {
  const params: Record<string, unknown> = {
    modelName: options.modelName,
    inOrderFields: options.fields,
    cardTemplates: options.cardTemplates,
    isCloze: false,
  };
  if (options.css !== undefined) {
    params.css = options.css;
  }
  return invoke("createModel", { params });
}

export async function addNotes(
  notes: Array<Record<string, unknown>>,
  options?: { timeoutMs?: number },
): Promise<Array<number | null>> {
  const result = await invoke("addNotes", {
    params: { notes },
    timeoutMs: options?.timeoutMs ?? 30_000,
  });
  if (!Array.isArray(result)) {
    throw new AnkiConnectError("AnkiConnect returned an invalid addNotes result.");
  }
  return result as Array<number | null>;
}

export async function syncWithAnkiWeb(options?: {
  timeoutMs?: number;
}): Promise<void> {
  await invoke("sync", { timeoutMs: options?.timeoutMs ?? 120_000 });
}

function quoteAnkiQueryValue(value: string): string {
  const escaped = value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
  return `"${escaped}"`;
}

export async function findNotes(
  query: string,
  options?: { timeoutMs?: number },
): Promise<number[]> {
  const result = await invoke("findNotes", {
    params: { query },
    timeoutMs: options?.timeoutMs ?? 30_000,
  });
  if (
    !Array.isArray(result) ||
    !result.every((noteId): noteId is number => typeof noteId === "number")
  ) {
    throw new AnkiConnectError("AnkiConnect returned an invalid findNotes result.");
  }
  return result;
}

export async function notesInfo(
  noteIds: number[],
  options?: { timeoutMs?: number },
): Promise<Array<Record<string, unknown>>> {
  if (noteIds.length === 0) {
    return [];
  }
  const result = await invoke("notesInfo", {
    params: { notes: noteIds },
    timeoutMs: options?.timeoutMs ?? 30_000,
  });
  if (
    !Array.isArray(result) ||
    !result.every(
      (item): item is Record<string, unknown> =>
        item !== null && typeof item === "object",
    )
  ) {
    throw new AnkiConnectError("AnkiConnect returned an invalid notesInfo result.");
  }
  return result;
}

export async function mappedNotesInDeck(
  deckName: string,
  logicalToAnki: Record<string, string>,
  options?: { timeoutMs?: number },
): Promise<Array<Record<string, string>>> {
  const query = `deck:${quoteAnkiQueryValue(deckName)}`;
  const noteIds = await findNotes(query, options);
  const notes: Array<Record<string, string>> = [];
  const chunkSize = 250;
  for (let start = 0; start < noteIds.length; start += chunkSize) {
    const chunk = noteIds.slice(start, start + chunkSize);
    for (const info of await notesInfo(chunk, options)) {
      const fields = info.fields;
      if (fields === null || typeof fields !== "object") {
        continue;
      }
      const mapped: Record<string, string> = {};
      for (const [logicalKey, ankiField] of Object.entries(logicalToAnki)) {
        const field = (fields as Record<string, unknown>)[ankiField];
        let raw: unknown;
        if (field !== null && typeof field === "object") {
          raw = (field as { value?: unknown }).value;
        } else {
          raw = field;
        }
        mapped[logicalKey] = typeof raw === "string" ? raw.trim() : "";
      }
      notes.push(mapped);
    }
  }
  return notes;
}

export async function isConnected(): Promise<boolean> {
  try {
    await deckNames();
    return true;
  } catch {
    return false;
  }
}
